### PR TITLE
Optimize Prometheus queries

### LINKED ISSUE #260: Optimize Prometheus queries
Certain Prometheus queries can become quite slow after Conduit has been running for some time --- in particular ones that use `histogram_quantile`.  We should optimize these queries so that they return quickly.

### PR DESCRIPTION
**Pull request by siggy**
Prometheus queries from the Telemetry service were taking seconds or 10s
of seconds.

Optimize these queries:
- Move all summary queries requiring a single data point off of Prometheus'
  QueryRange() endpoint, onto Query()
- Set `defaultVectorRange` to 30s, and also use it regardless of time
  window
Also add tests for grpc_server and telemetry server

Signed-off-by: Andrew Seigner <siggy@buoyant.io>

Fixes #260

# Results

Previously we had 12 variants of Prometheus query with p50's ranging from 4s to 8s. After these changes, all queries are < 3s, and all but 9 are < 300ms. p95 and p99 are also cut in half.

All results were generated using https://github.com/runconduit/conduit/tree/siggy/viz

### Before

sum_avg_baseline

### After

screen shot 2018-02-07 at 10 15 26 am

### Before

avg_baseline_step

### After

avg_query_step

### CODE REVIEW

File: controller/telemetry/server.go [hunk 1]
```diff
@@ -184,29 +184,52 @@ func NewServer(addr, prometheusUrl string, ignoredNamespaces []string, kubeconfi
 func (s *server) Query(ctx context.Context, req *read.QueryRequest) (*read.QueryResponse, error) {
```
Comment (adleong) [inline controller/telemetry/server.go]: Much of the behavior of this method switches depending on if start/stop/step is set or not.  Would it make sense to split this into two methods, `QueryTs` and `QuerySummary` or something like that?

File: controller/telemetry/server.go [hunk 2]
```diff
@@ -184,29 +184,52 @@ func NewServer(addr, prometheusUrl string, ignoredNamespaces []string, kubeconfi
...
 
-	if res.Type() != model.ValMatrix {
-		return nil, fmt.Errorf("Unexpected query result type: %s", res.Type())
+		res, err = s.prometheusAPI.QueryRange(ctx, req.Query, queryRange)
+		if err != nil {
+			log.Errorf("QueryRange(%+v, %+v) failed with: %+v", req.Query, queryRange, err)
+			return nil, err
+		}
+	} else {
+		res, err = s.prometheusAPI.Query(ctx, req.Query, time.Time{})
```
Comment (adleong) [inline controller/telemetry/server.go]: This is "summary" mode?  Does that means this gets the single most recent datapoint or does it summarize over some range?

File: controller/telemetry/server.go [hunk 3]
```diff
@@ -184,29 +184,52 @@ func NewServer(addr, prometheusUrl string, ignoredNamespaces []string, kubeconfi
...
+		if err != nil {
+			log.Errorf("Query(%+v, %+v) failed with: %+v", req.Query, time.Time{}, err)
+			return nil, err
+		}
 	}
+	log.Debugf("Query response: %+v", res)
 
 	samples := make([]*read.Sample, 0)
-	for _, s := range res.(model.Matrix) {
-		samples = append(samples, convertSampleStream(s))
```
Comment (adleong) [inline controller/telemetry/server.go]: this is not specific to this change, but I'm struggling a bit to understand the API.  Is a "Sample" a timeseries?  and each sample contains the data points for that timeseries?  and in summary mode each Sample will only have one data point?

File: controller/api/public/grpc_server_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	mReq *pb.MetricRequest
+	mRes *pb.MetricResponse
+}
+
+func TestStat(t *testing.T) {
+	t.Run("Stat returns the expected responses", func(t *testing.T) {
+
+		responses := []testResponse{
+			testResponse{
+				tReq: &telemetry.QueryResponse{
```
Comment (rmars) [inline controller/api/public/grpc_server_test.go]: should this variable name be `tRes`?

File: controller/telemetry/server_test.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		req := &read.QueryRequest{
+			Query: "fake query",
+		}
+		_, err := s.Query(context.Background(), req)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+	})
+
+	t.Run("Query with time range performs timeseries request", func(t *testing.T) {
```
Comment (rmars) [inline controller/telemetry/server_test.go]: I might be missing something, but this test (and the one in the next section) passes even if you don't include a time range. are we just testing that it accepts such an input without error?

File: controller/telemetry/server.go (hunk 1 above)
Comment (siggy) [reply controller/telemetry/server.go]: I agree, now that the telemetry service's `Query()` endpoint is supporting two distinct use cases, this warrants modifying the API to be more explicit about it. This will require a protobuf change in the telemetry service, which is better done in a separate PR, I've filed #312 to track.

File: controller/telemetry/server.go (hunk 2 above)
Comment (siggy) [reply controller/telemetry/server.go]: Correct this is "summary" mode, which means return the single most recent data point. We pass in `time.Time{}` as a way to indicate we want the most recent data point:
https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries

We were previously doing this manually by querying the full timeseries, and then calling `filterQueryRsp` to filter to the latest data point.

For a bit more context, here is where the `timeseries` param in the `web` API translates into `Summarize` in the public API (perhaps as part of #312 we could make these terms more consistent):
https://github.com/runconduit/conduit/blob/2f114e69fad36c7eb40aefcbccf301a6c30470c6/web/srv/api_handlers.go#L140

File: controller/telemetry/server.go (hunk 3 above)
Comment (siggy) [reply controller/telemetry/server.go]: correct, a `read.Sample` represents a single metric over time, and, in summary mode `read.Sample` will contain one data point. this is another artifact of telemetry's `Query()` endpoint supporting both timeseries and non-timeseries data (so it always returns a timeseries-like data structure). this is another area to clean up in #312.

File: controller/telemetry/server_test.go (hunk 5 above)
Comment (siggy) [reply controller/telemetry/server_test.go]: Really good catch !

I had previously implemented telemetry service's `Query()` endpoint to handle all querying first, then all data processes second. I felt these edge cases were worth the trade off to make the code a bit easier to read. Following your comment here, I've reconsidered, and have reorganized `Query()` to handle timeseries request/process in one block, and non-timeseries in another. This produces slightly more repeated code, but as you've pointed out, is more correct. I've modified the tests as well to validate these additional edge cases. Thanks!

File: controller/telemetry/server_test.go [hunk 6]
```diff
@@ -152,39 +143,20 @@ func TestServerResponses(t *testing.T) {
...
-				Query: "fake query",
-			}
-			res, err := s.Query(context.Background(), req)
-			if err != nil {
-				t.Fatalf("Unexpected error: %v", err)
+			res, err := s.Query(context.Background(), tr.queryReq)
+			if err != nil || tr.err != nil {
+				if (err == nil && tr.err != nil) ||
+					(err != nil && tr.err == nil) ||
+					(err.Error() != tr.err.Error()) {
```
Comment (rmars) [inline controller/telemetry/server_test.go]: omggg
